File: desktop/CredentialTyper.Transport/Wire/QrPayload.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace CredentialTyper.Transport.Wire;

/// <summary>
/// 二维码内容：桌面身份公钥 + LAN 地址 + 一次性配对码。
/// 序列化为 UTF-8 JSON 再 base64url，塞进二维码。
/// </summary>
public sealed class QrPayload
{
    [JsonPropertyName("v")] public int Version { get; set; } = Protocol.Version;

    /// <summary>桌面静态公钥（32 字节）base64url。</summary>
    [JsonPropertyName("spub")] public string SPub { get; set; } = "";

    /// <summary>桌面所有 LAN IPv4，手机逐个尝试。</summary>
    [JsonPropertyName("host")] public string[] Host { get; set; } = Array.Empty<string>();

    [JsonPropertyName("port")] public int Port { get; set; }

    /// <summary>一次性配对码（16 字节）base64url。</summary>
    [JsonPropertyName("code")] public string Code { get; set; } = "";

    [JsonIgnore] public byte[] SPubBytes => Base64Url.Decode(SPub);
    [JsonIgnore] public byte[] CodeBytes => Base64Url.Decode(Code);

    private static readonly JsonSerializerOptions Opts = new() { DefaultIgnoreCondition = JsonIgnoreCondition.Never };

    public static QrPayload Create(byte[] spub, string[] host, int port, byte[] code) => new()
    {
        SPub = Base64Url.Encode(spub),
        Host = host,
        Port = port,
        Code = Base64Url.Encode(code),
    };

    /// <summary>整体 base64url，即二维码里的字符串。</summary>
    public string ToQrText() => Base64Url.Encode(JsonSerializer.SerializeToUtf8Bytes(this, Opts));

    public static QrPayload FromQrText(string qr)
    {
        var json = Base64Url.Decode(qr);
        return JsonSerializer.Deserialize<QrPayload>(json, Opts)
            ?? throw new FormatException("二维码内容无法解析为 QrPayload");
    }

    public string ToJson() => Encoding.UTF8.GetString(JsonSerializer.SerializeToUtf8Bytes(this, Opts));
}

File: app/lib/noise/qr_payload.dart
import 'dart:convert';
import 'dart:typed_data';

import 'base64url.dart';
import 'protocol.dart';

/// 二维码内容：桌面身份公钥 + LAN 地址 + 一次性配对码。
/// 序列化为 UTF-8 JSON 再 base64url。对应桌面 `QrPayload.cs`。
class QrPayload {
  final int version;
  final String sPub; // 桌面静态公钥 base64url
  final List<String> host; // 桌面所有 LAN IPv4
  final int port;
  final String code; // 一次性配对码 base64url

  const QrPayload({
    required this.version,
    required this.sPub,
    required this.host,
    required this.port,
    required this.code,
  });

  Uint8List get sPubBytes => Base64Url.decode(sPub);
  Uint8List get codeBytes => Base64Url.decode(code);

  static QrPayload fromQrText(String qr) {
    // QR scanners and clipboard managers may add line breaks, a BOM, or
    // zero-width spaces.  They are not part of the base64url payload.
    final cleaned = qr.replaceAll(
      RegExp(r'[\u0000-\u0020\u00a0\u200b\u200c\u200d\ufeff]'),
      '',
    );
    if (cleaned.isEmpty) throw const FormatException('连接码为空');

    late final Map<String, dynamic> map;
    try {
      final jsonBytes = Base64Url.decode(cleaned);
      final decoded = jsonDecode(utf8.decode(jsonBytes));
      if (decoded is! Map<String, dynamic>) {
        throw const FormatException('连接码格式错误');
      }
      map = decoded;
    } on FormatException {
      rethrow;
    } catch (_) {
      throw const FormatException('连接码格式错误');
    }

    final version = (map['v'] as num?)?.toInt() ?? 0;
    final sPub = map['spub'] as String? ?? '';
    final code = map['code'] as String? ?? '';
    final hosts = (map['host'] as List<dynamic>? ?? const [])
        .map((e) => e.toString().trim())
        .where((e) => e.isNotEmpty)
        .toSet()
        .toList();
    final port = (map['port'] as num?)?.toInt() ?? 0;

    if (version != Protocol.version) {
      throw FormatException('不支持的连接码版本：$version');
    }
    if (hosts.isEmpty) throw const FormatException('连接码缺少电脑地址');
    if (port < 1 || port > 65535) throw const FormatException('连接码端口无效');

    try {
      if (Base64Url.decode(sPub).length != Protocol.publicKeyLen ||
          Base64Url.decode(code).length != Protocol.pairingCodeLen) {
        throw const FormatException('连接码指纹或配对码无效');
      }
    } catch (_) {
      throw const FormatException('连接码指纹或配对码无效');
    }

    return QrPayload(
      version: version,
      sPub: sPub,
      host: hosts,
      port: port,
      code: code,
    );
  }

  String toQrText() {
    final json = jsonEncode({
      'v': version,
      'spub': sPub,
      'host': host,
      'port': port,
      'code': code,
    });
    return Base64Url.encode(utf8.encode(json));
  }
}
